Sources/NIOCore/AsyncSequences/NIOAsyncWriter.swift

```diff
@@ -0,0 +1,789 @@
+    /// right away to the delegate. If the ``NIOAsyncWriter`` was _NOT_ writable then the sequence will be buffered
+    /// until the ``NIOAsyncWriter`` becomes writable again.
+    func didYield<S: Sequence>(contentsOf sequence: S) where S.Element == Element
+
+    /// This method is called once the ``NIOAsyncWriter`` is terminated.
+    ///
+    /// Termination happens if:
+    /// - The ``NIOAsyncWriter`` is deinited.
+    /// - ``NIOAsyncWriter/finish()`` is called.
+    /// - ``NIOAsyncWriter/finish(with:)`` is called.
+    ///
+    /// - Note: This is guaranteed to be called _exactly_ once.
+    ///
+    /// - Parameter failure: The failure that terminated the ``NIOAsyncWriter``. If the writer was terminated without an
+    /// error this value is `nil`.
+    func didTerminate(failure: Failure?)
+}
+
+/// Errors thrown by the ``NIOAsyncWriter``.
+@available(macOS 10.15, iOS 13, tvOS 13, watchOS 6, *)
+public struct NIOAsyncWriterError: Error, Hashable {
+    @usableFromInline
+    internal enum _Code: Hashable, Sendable {
+        case alreadyFinished
+    }
+
+    @usableFromInline
+    let _code: _Code
+
+    @usableFromInline
+    var file: String
+
+    @usableFromInline
+    var line: Int
+
+    @inlinable
+    init(_code: _Code, file: String = #fileID, line: Int = #line) {
+        self._code = _code
+        self.file = file
+        self.line = line
+    }
+
+    public static func == (lhs: NIOAsyncWriterError, rhs: NIOAsyncWriterError) -> Bool {
+        return lhs._code == rhs._code
+    }
+
+    public func hash(into hasher: inout Hasher) {
+        hasher.combine(self._code)
+    }
+
+    /// Indicates that the ``NIOAsyncWriter`` has already finished and is not accepting any more writes.
+    public static let alreadyFinished: Self = .init(_code: .alreadyFinished)
+}
+
+/// A ``NIOAsyncWriter`` is a type used to bridge elements from the Swift Concurrency domain into
+/// a synchronous world. The `Task`s that are yielding to the ``NIOAsyncWriter`` are the producers.
+/// Whereas the ``NIOAsyncWriterDelegate`` is the consumer.
+///
+/// Additionally, the ``NIOAsyncWriter`` allows the consumer to toggle the writability by calling ``NIOAsyncWriter/Sink/toggleWritability()``.
+/// This allows the implementation of flow control on the consumer side. Any call to ``NIOAsyncWriter/yield(contentsOf:)`` or ``NIOAsyncWriter/yield(_:)``
+/// will suspend if the ``NIOAsyncWriter`` is not writable and only be resumed after the ``NIOAsyncWriter`` becomes writable again.
+///
+/// - Note: It is recommended to never directly expose this type from APIs, but rather wrap it. This is due to the fact that
+/// this type has three generic parameters where at least two should be known statically and it is really awkward to spell out this type.
+/// Moreover, having a wrapping type allows to optimize this to specialized calls if all generic types are known.
+@available(macOS 10.15, iOS 13, tvOS 13, watchOS 6, *)
+public struct NIOAsyncWriter<
```

Comment:
I'm not sure, I'm a fan of this being a struct. The AsyncWriter has clearly reference semantics.

Reply:
Fair point, I think making this a class isn't bad. I went with the struct first approach but open to change it